pkg/provisioner/ironic/ironic.go

```diff
@@ -795,9 +793,69 @@ func (p *ironicProvisioner) getUpdateOptsForNode(ironicNode *nodes.Node) (update
 		},
 	)
 
+	// boot_mode
+	op, value := buildCapabilitiesValue(ironicNode, p.host.Status.Provisioning.BootMode)
+	updates = append(
+		updates,
+		nodes.UpdateOperation{
+			Op:    op,
+			Path:  "/properties/capabilities",
+			Value: value,
+		},
+	)
+
 	return updates, nil
 }
 
+// We can't just replace the capabilities because we need to keep the
+// values provided by inspection. We can't replace only the boot_mode
+// because the API isn't fine-grained enough for that. So we have to
+// look at the existing value and modify it. This function
+// encapsulates the logic for building the value and knowing which
+// update operation to use with the results.
+func buildCapabilitiesValue(ironicNode *nodes.Node, bootMode metal3v1alpha1.BootMode) (op nodes.UpdateOp, value string) {
+
+	capabilities, ok := ironicNode.Properties["capabilities"]
+	if !ok {
+		// There is no existing capabilities value
+		return nodes.AddOp, bootModeCapabilities[bootMode]
+	}
+	existingCapabilities := capabilities.(string)
+
+	// The capabilities value is set, so we will want to replace it.
+	op = nodes.ReplaceOp
+
+	if existingCapabilities == "" {
+		// The existing value is empty so we can replace the whole
+		// thing.
+		value = bootModeCapabilities[bootMode]
+		return
+	}
+
+	if !strings.Contains(existingCapabilities, "boot_mode") {
+		// No boot_mode is set, append and return
+		value = fmt.Sprintf("%s,%s", existingCapabilities, bootModeCapabilities[bootMode])
+		return
+	}
+
+	// The capabilities value has format "var1:val1,var2:val2". We
+	// know that boot_mode is there but not what value it has.  There
+	// are only 2 values, so we can replace the "wrong" string with
+	// the right one without fully parsing the string. We may want to
+	// change this later when we have more boot modes.
```

Comment:
I hope we're going to have less boot modes over time, not more :) On the other hand, you cannot fully rule out another capability having "uefi" in its name. Maybe search and replace the whole substring boot_mode:uefi/boot_mode:bios?

Or you can cargo-cult the whole parsing from https://opendev.org/openstack/ironic/src/branch/master/ironic/common/utils.py#L401

Reply:
Yes `bootModeCapabilities[fromMode]` will include `boot_mode:` and then the right mode, so we at least get the full string.

It would be safer to do the full semantic parsing of the input, but if we add another boot mode with `uefi` in the name we're going to need to touch a bunch of this stuff to update the enum anyway, so we can do that work then.

Reply:
Ah, I didn't realize that.